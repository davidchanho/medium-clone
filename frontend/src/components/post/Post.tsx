import React from "react";
import { clampText } from "../../helpers/clampText";
import { formatDate } from "../../helpers/formatDate";
import { IPost } from "../../types";
import PostOptions from "../post-options";

function Post({ ...post }: IPost) {
  const { publication, user, title, body, excerpt, readingTime, date } = post;

  return (
    <div>
      <div className="d-flex align-items-center">
        <img
          src={publication?.icon}
          alt={publication?.name}
          className="mr-2 mt-1 rounded-circle"
          width="20"
          height="20"
        />{" "}
        <p>
          {user?.name} in {publication?.name}
        </p>
      </div>
      <div className="text-secondary d-flex flex-column">
        <h5 className="text-capitalize mb-1">{title}</h5>
        <p>{excerpt ? clampText(body, excerpt) : ""}</p>
        <div className="d-flex align-items-center justify-content-between">
          <p>
            {formatDate(date)} &middot; {readingTime}
          </p>
          <PostOptions {...post} />
        </div>
      </div>
    </div>
  );
}

export default Post;
