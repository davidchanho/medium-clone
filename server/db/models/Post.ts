import { Document, model, Schema } from "mongoose";
import { ICommentDoc } from "./Comment";
import { IPublicationDoc } from "./Publication";
import { IUserDoc } from "./User";

export interface IPostDoc extends Document {
  publication: IPublicationDoc;
  user: IUserDoc;
  title: string;
  body: string;
  image: string;
  readingTime: string;
  date: string;
  claps: number;
  comments: ICommentDoc[];
}

const postSchema = new Schema(
  {
    publication: {
      type: Schema.Types.ObjectId,
      ref: "publication",
    },
    user: { type: Schema.Types.ObjectId, ref: "user" },
    title: {
      type: String,
      minlength: [6, "post title must be at least 6 characters long"],
      required: [true, "post title is required"],
    },
    body: {
      type: String,
      minlength: [6, "post body must be at least 6 characters long"],
      required: [true, "post body is required"],
    },
    image: {
      type: String,
      default:
        "https://miro.medium.com/fit/c/272/181/1*ts3kXw7MPHS_yrejwIbZHA.png",
    },
    readingTime: {
      type: String,
    },
    claps: { type: Number, default: 0 },
    comments: [{ type: Schema.Types.ObjectId, ref: "comment" }],
    date: {
      type: String,
      default: new Date().toString(),
    },
  },
  { timestamps: true }
);

const Post = model<IPostDoc>("post", postSchema);

export default Post;
