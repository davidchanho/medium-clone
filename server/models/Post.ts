import { Document, model, Schema } from "mongoose";

export interface IPostDoc extends Document {
  title: string;
  body: string;
}

const postSchema = new Schema(
  {
    title: {
      type: String,
      minlength: [6, "post title must be at least 6 characters long"],
      required: [true, "post title is required"],
    },
    body: {
      type: String,
      minlength: [6, "post body must be at least 6 characters long"],
      required: [true, "post body is required"],
    },
  },
  { timestamps: true }
);

const Post = model<IPostDoc>("post", postSchema);

export default Post;
