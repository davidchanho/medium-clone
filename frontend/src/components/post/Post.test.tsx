import React from "react";
import { BrowserRouter as Router} from "react-router-dom";
import renderer from "react-test-renderer";
import Post from ".";
import { testPost } from "../../tests";

describe("Post", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Router>
          <Post {...testPost} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
