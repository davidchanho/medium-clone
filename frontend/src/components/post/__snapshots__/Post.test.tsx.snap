// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post renders correctly 1`] = `
<div
  className="sc-bdvvaa"
  onClick={[Function]}
>
  <header
    className="sc-gsDJrp dSpPBI"
  >
    <img
      alt="test"
      className="sc-dkPtyc gFABCv"
      height="20"
      src="test"
      width="20"
    />
     
    <p>
      test
       in 
      test
    </p>
  </header>
  <div>
    test
  </div>
  <div>
    
    Invalid Date
     Â· 
    test
  </div>
</div>
`;
